heu/library/algorithms/dgk/key_generator.cc

```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 zhangwfjh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "heu/library/algorithms/dgk/key_generator.h"
+
+#include "yacl/base/exception.h"
+
+#include "heu/library/algorithms/util/mp_int.h"
+
+namespace heu::lib::algorithms::dgk {
+
+void KeyGenerator::Generate(size_t key_size, SecretKey* sk, PublicKey* pk) {
+  YACL_ENFORCE(key_size % 2 == 0, "Key size must be even");
+  YACL_ENFORCE(key_size >= 1024 && key_size <= 3072,
+               "Key size must be in [1024, 3072] and 2048 bits is recommended");
+
+  MPInt u, vp, vq;
+  // MPINT_ENFORCE_OK(
+  //     mp_prime_rand(reinterpret_cast<mp_int*>(&u), 1, l, MP_PRIME_BBS));
+  u = MPInt{65423};  // use the largest l(16)-bit prime instead
+  MPInt::RandPrimeOver(t, &vp);
+  MPInt::RandPrimeOver(t, &vq);
+  MPInt wp, wq, p, q, gcd;
+  do {
+    MPInt::RandomMonicExactBits(key_size / 2 - t - l, &wp);
+    MPInt::Gcd(wp, vq, &gcd);
+    wp *= MPInt::_2_;
```

Comment:
这里方便给一下这种计算方法的出处嘛？

Reply:
To obtain $m$, we compute $c^{v_p} \equiv g^{v_pm}h^{v_pr} \bmod p$. Since $h^{v_pv_q} \equiv 1\bmod n$, we have $h^{v_pv_q} \equiv 1\bmod p$. It follows from $\gcd(v_q, p-1) = \gcd(v_q, 2v_pw_p)=\gcd(v_q, w_p)=1$ that $h^{v_p} \equiv 1 \bmod p$.

Reply:
LGTM; FYI 这里可以帮忙加一下 下面的 comments 吗？
```cpp
// Question: can we consider the following generations of p, q secure? 
// TODO: check NIST.FIPS.186-5 Appendix A.1.1
```

Reply:
done